Multi node config

fabriziopandini opened the pull request:
This PR completes the implementation of support for multi node configurations in Kind, leveraging on changes introduced by https://github.com/kubernetes-sigs/kind/pull/137 and https://github.com/kubernetes-sigs/kind/pull/143

fixes: https://github.com/kubernetes-sigs/kind/issues/130

**Note for reviewers** 
While implementing I tried to keep the kind Config surface as minimal as possible and now the `v1alpha2` version of kind configuration has only one object called `Node`, that can be repeated many times in a yaml document and/or set to automatically generate replicas. e.g.

```yaml
# ha cluster with 3 control-plane nodes, 2 workers and an external load balancer
kind: Node
apiVersion: kind.sigs.k8s.io/v1alpha2
role: control-plane
replicas: 3
---
kind: Node
apiVersion: kind.sigs.k8s.io/v1alpha2
role: worker
replicas: 2
---
kind: Node
apiVersion: kind.sigs.k8s.io/v1alpha2
role: external-load-balancer
```

Nodes are then grouped in a `Config` object, but this object exists only in the  internal version of kind configuration and so the user should not care about it.

Automatic conversion from `v1alpha1` configurations is supported.

Finally, please note that while the `v1alpha2` version of kind configuration supports multi-node, the rest of Kind still not, so the user will be temporarily blocked when trying to create clusters with n>1 nodes.

neolit123 commented on pkg/cluster/config/config_types_test.go:
```diff
@@ -0,0 +1,189 @@
+
+func checkNode(t *testing.T, n *Node, name *string) {
+	if (n == nil) != (name == nil) {
+		t.Errorf("expected %v node, saw %v", name, n)
```
isn't it the same as doing `n != name` ?

also not sure how to formulate this message.
ideally we should be comparing objects of the same type e.g. bool vs bool.

fabriziopandini replied:
n != name can't be used because n and name are different types

the test is validate node ordering, and checking only the node name is a way to do this and at the same time keep test case definition simple and readable

neolit123 commented on pkg/cluster/config/encoding/scheme.go:
```diff
@@ -43,59 +49,172 @@ func init() {
+// can be one of the different API versions defined in scheme.
+// If path == "" then the default config is returned
+func Load(path string) (*config.Config, error) {
+	if path == "" {
```
are there messages in the pipeline saying "using default config path" vs "using **your** config path"

fabriziopandini replied:
The only use case where "your config path" will be used, is when the user pass `--config` explicitly, so he/she is already aware of what's going on...

neolit123 commented on pkg/cluster/config/encoding/scheme.go:
```diff
@@ -43,59 +49,172 @@ func init() {
+// one of the different API versions defined in the Scheme; for v1alpha2
+// multiple yaml documents within the same yaml file are supported
+func unmarshalConfig(contents []byte) (*config.Config, error) {
+	// parses yamlDocuments separated by --- directives
```
so basically multiple objects of kind Node at this point?

neolit123 commented on pkg/cluster/config/config_types.go:
```diff
@@ -0,0 +1,220 @@
+		// node with external etcd role
+		if replica.IsExternalEtcd() {
+			if c.externalEtcd != nil {
+				return fmt.Errorf("invalid config. there are two nodes with role %q", ExternalEtcdRole)
```
we should use the pkg/errors package everywhere.
https://godoc.org/github.com/pkg/errors

neolit123 commented on pkg/cluster/config/encoding/scheme.go:
```diff
@@ -43,59 +49,172 @@ func init() {
+
+// splitYAMLDocuments divides a yaml file into yamlDocuments,
+// and identifies each document with its own GroupVersionKind
+func splitYAMLDocuments(contents []byte) (yamlDocuments []yamlDocument, err error) {
```
we really need a unified location for this type of utility.
this is pretty much duplicating kubeadm at this point. :\

and once the ComponentConfig KEP stuff is in place we need to revise all locations that bake their own utility.

neolit123 commented on pkg/cluster/config/encoding/scheme_test.go:
```diff
@@ -17,66 +17,213 @@ limitations under the License.
 
-// TODO(fabrizio pandini): once we have multiple config API versions we
-// will need more tests
+func TestSplitYAMLDocuments(t *testing.T) {
```
and if we have a unified location for the split utility we can avoid some of the tests too.

neolit123 commented on pkg/cluster/config/types.go:
```diff
@@ -75,3 +108,52 @@ type LifecycleHook struct {
 }
+
+// IsControlPlane returns true if the node hosts a control plane instance
+func (n *Node) IsControlPlane() bool {
```
this is something that i've mentioned earlier - are public methods for root internal config objects something we want?

neolit123 commented on pkg/cluster/config/v1alpha2/doc.go:
```diff
@@ -0,0 +1,23 @@
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=sigs.k8s.io/kind/pkg/cluster/config
+// +k8s:defaulter-gen=TypeMeta
+package v1alpha2
```
maybe it's time to start adding docs with Node config examples.

BenTheElder replied:
yes, let's document this once it's ready for use, and until then keep it to the config example file, for those that want to do some digging.

neolit123 commented on pkg/cluster/config/v1alpha2/types.go:
```diff
@@ -0,0 +1,101 @@
+	KubeadmConfigPatches []string `json:"kubeadmConfigPatches,omitempty"`
+	// KubeadmConfigPatchesJSON6902 are applied to the generated kubeadm config
+	// as patchesJson6902 to `kustomize build`
+	KubeadmConfigPatchesJSON6902 []kustomize.PatchJSON6902 `json:"kubeadmConfigPatchesJson6902,omitempty"`
```
side note: i'd be interested to hear user feedback on `KubeadmConfigPatches` and `KubeadmConfigPatchesJSON6902`

BenTheElder replied:
as possibly the sole user currently:
- it works ish
- sometimes for advanced field patching the 6902 format is more powerful, but the plain yaml patch style is much nicer to write and very simple to understand, it just merges the yaml kinda like apply.
- it's not the _best_ UX, but you also can ignore it and isntead run config patching in a lifecycle hook, though the UX will be worse for that.

neolit123 commented on pkg/cluster/config/validate.go:
```diff
@@ -26,12 +26,53 @@ import (
+	// All nodes in the config should be valid
+	for _, n := range c.nodes {
+		if err := n.Validate(); err != nil {
+			errs = append(errs, fmt.Errorf("Please fix invalid configuration for node %q: \n%v", n.Name, err))
```
errors in this file seem to start with uppercase.
also we should use pkg/errors.

fabriziopandini replied:
fixed

neolit123 commented on pkg/cluster/config/validate_test.go:
```diff
@@ -93,39 +94,137 @@ func TestConfigValidate(t *testing.T) {
+			// Adding nodes to the config
+			for _, n := range tc.Nodes {
+				if err := c.Add(n); err != nil {
+					t.Fatalf("Unexpected error while adding nodes: %v", err)
```
lowercase error.

fabriziopandini replied:
fixed

neolit123 commented on pkg/cluster/context.go:
```diff
@@ -138,15 +138,19 @@ func (c *Context) Create(cfg *config.Config, retain bool) error {
-	image := cfg.Image
+
+	// TODO(fabrizio pandini): usage of BootStrapControlPlane() is temporary / WIP
+	// kind v1alpha config fully supports multi nodes, but the cluster creation logic implemented in
```
`v1alpha2 config`

fabriziopandini replied:
fixed

BenTheElder commented on pkg/cluster/config/config_types.go:
```diff
@@ -0,0 +1,220 @@
+/*
```
:+1: I'd kinda prefer that the config stay just data and we put logic elsewhere other than like, copy, convert. @munnerz voiced a similar opinion.

OTOH, a few of those things are public methods like DeepCopy

fabriziopandini replied:
@BenTheElder @neolit123 @munnerz 
There are drawbacks in what you are asking

`Config` in v1alpha2 is an helper object that wraps many nodes. It is defined only in the internal version of the config, and it is not available in the public api.

By having this helper object nears few utilities functions I can keep `Config` fields private, and thus ensure that `Nodes` and all the info derived by Nodes respect a set of assumptions that  will greatly simplifies the rest of the code. e.g. Nodes are ordered by provisioning order, Node names are unique, info derived from node are properly set (e.g. bootstrap control plane identified in a consistent way etc)

A prior art for this approach is [`Scheme`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/scheme.go): it is used only internally, it acts as a container for other types, it has private fields and public func for handling internal state.

If we want to split Config by its helper functions probably I will be forced to make `Config` fields public, and then loose some/all the above guarantees. As a consequence all the rest of the code IMO will result much more fragile/error prone.

There are other notable prior arts for where helper function are kept near types e.g.:
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/meta.go
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/helpers.go

Said that I'm open to discuss any suggestion helps in cleaning config while preserving above guarantees.

In the meantime, I tried to make things more clear and now types and helper function are in separated files. Also comments were improved accordingly

neolit123 replied:
some random points:

> Config in v1alpha2 is an helper object that wraps many nodes. It is defined only in the internal version of the config, and it is not available in the public api.

i guess we are also talking about the Node object, that is public and has it's fields public.
so we don't have to have public methods atteched to the Node struct, but we can instead use helper functions that accept a *Node as parameter.

> https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/meta.go

the metav1 is a good comparison of an API that has public methods for the config struct, i guess.

i was thinking of a package that is importable and has types and interfaces but no definitions. but golang smears the line between declarations and definitions.

leaving it to @BenTheElder for the final call on the public methods.

BenTheElder replied:
this seems fairly easy to re-organize before V1, but I would tend to agree that config should be plain old data and helpers can accept a config object.

fabriziopandini commented:
@BenTheElder 
Rebased after merge of #143, addressed minor comments, answered to other comments

@neolit123 
> i think we should also expose a way to have a common configuration... i wonder if we can have something like a ConfigMap

IMO providing a shortcut for configuration changes is part of #133.
However, it is possible to iterate on this also in future because, if possible, I prefer to keep this PR small and focused.

> how about multi-cluster?

As far as I understand multi cluster is supported in `kind` by decoupling the cluster name from the config. This is preserved by this PR, so you can have multi-cluster/multi-nodes

neolit123 commented:
> As far as I understand multi cluster is supported in kind by decoupling the cluster name from the config. This is preserved by this PR, so you can have multi-cluster/multi-nodes

using the already existing decoupling, would it be possible to configure a multi-cluster topology where an external load balancer serves in front of both clusters?

thanks for the updates.
/lgtm 
/hold

fabriziopandini commented:
> would it be possible to configure a multi-cluster topology where an external load balancer serves in front of both clusters?

In my hacky prototype I'm creating an ha proxy for each cluster, and I think I will start from this assumption also in moving this feature upstream because implementation will be isimpler without cross cluster dependencies.

rdodev commented:
This looks really great @fabriziopandini looking forward to take it for a spin.

BenTheElder commented:
> using the already existing decoupling, would it be possible to configure a multi-cluster topology where an external load balancer serves in front of both clusters?

related: https://github.com/kubernetes-sigs/federation-v2 is doing some work with `kind`, we should speak to them.

https://github.com/kubernetes-sigs/federation-v2/blob/c8398aa4f6439e4e0603c75e513d394a9b4f2595/scripts/download-e2e-binaries.sh#L36-L45

BenTheElder commented on pkg/cluster/config/v1alpha2/zz_generated.conversion.go:
```diff
@@ -149,6 +123,38 @@ func Convert_config_LifecycleHook_To_v1alpha2_LifecycleHook(in *config.Lifecycle
+	out.Replicas = (*int32)(unsafe.Pointer(in.Replicas))
+	out.Role = NodeRole(in.Role)
+	out.Image = in.Image
+	out.KubeadmConfigPatches = *(*[]string)(unsafe.Pointer(&in.KubeadmConfigPatches))
```
ok thanks, I wonder why they're doing this so much.

fabriziopandini replied:
I can only guess:
- because conversion-gen always assumes that we are converting from TypeA.vx to TypeA.vY and does not detect cases when the type is external from the API (the type is the same)
or
- because this is a convenient way for doing "deepcopy" with types not supporting it explicitly

BenTheElder commented on pkg/cluster/config/helpers.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+// IsWorker returns true if the node hosts a worker instance
+func (n *Node) IsWorker() bool {
```
thanks!

fabriziopandini commented:
@neolit123 if I'm not wrong all the comment s are addressed.
Could you kindly leave hold?

neolit123 commented:
thanks for the updates @fabriziopandini
/hold cancel

neolit123 commented:
the more i think about it the more i prefer the node list under a single config parent object:
```yaml
kind: KindConfig
apiVersion: kind.sigs.k8s.io/v1alpha2
nodes:
- role: control-plane
  replicas: 3
- role: worker
  replicas: 2
- role: external-load-balancer
```

i see this method as easier to "UX" and maintain.
maybe i'm missing the benefits of doing the separate GVK for Node objects?

fabriziopandini commented:
@neolit123 
There are pro and cons. A list is not ideal when you have only one node, and a list somehow constraints you in having all the elements of a list equal, which I'm note sure is something we want in the long term (e.g. having hooks for external load balancer and etcd does not meek entirely sense to me)

Said that, I think that at this stage it is really important to go out as soon as possible in order to start getting using feedback and unblock the remaining part of multi-node activities (some new requirements for config can be identified along the way as well)

BenTheElder commented:
I finally got some time to just think about this some more and I think the nodes should be a list in a top level object. like @neolit123 pointed out

- This gives us room to add fields that are config wide, rather than on each node
- this ensures that nodes do not have distinct GVK (!), which would otherwise be problematic to support well.

BenTheElder commented:
Apologies for the delay on this PR, KubeCon and then post-con "plague" (a cold?) built up a pretty large backlog of reviews. Will be getting to these more regularly now.

fabriziopandini commented:
@BenTheElder I hope you are well now!

I'm going to address comments because I think it is important to get a first release merged ASAP to get moving also following parts of multi-node efforts, but IMO this leads to a more fragile design, but this is not blocking because we can always iterate in future.

Those are the point that IMO will be the candidate for future iterations:
- having "config wide" fields is an orthogonal problem to having a list/a sequence of objects (in any case "config wide" fields should be hosted on a different object than the list itself)
- having "config wide" fields opens up to precedence/override problems on corresponding node fields (if any)
- having a list assumes all the node are equal, which is a wrong assumption in multi-node/HA scenario
- both the kubeadm config (after several reworks) and the cluster API are implementing different approaches

BenTheElder commented:
Feeling a bit better, thanks :-)

Note: I do *not* want to rush multi-node _too_ much, we are already in need to removing some technical debt from kind. Previous iterations had multi-node and were not right.

These are interesting points but I'm not convinced.

> having "config wide" fields is an orthogonal problem to having a list/a sequence of objects (in any case "config wide" fields should be hosted on a different object than the list itself)

The list would become a subset. The reason for doing this is to always give the rest of the code a "Config" object to work with, which should be as much as possible simply something read from disk and desearialized with minimal mutation (just defaulting).

> having "config wide" fields opens up to precedence/override problems on corresponding node fields (if any)

The Node fields take precedence over the config wide if any of them somehow clashed. They generally shouldn't though. The same is said in k8s of say resource requests vs inline requests.

> having a list assumes all the node are equal, which is a wrong assumption in multi-node/HA scenario

It only assumes there is a list of "Node" type - why wouldn't there be? Nodes can have different roles.

> both the kubeadm config (after several reworks) and the cluster API are implementing different approaches

I don't think the cluster API in it's current state is desirable to emulate for local clusters. Kubeadm only has separate objects due to the phases. We're not exposing that. There's also no real reason these can't be combined into a single object, I don't think we want the burden of nodes with different GVK.

In general elevating individual node specs to have GVK seems very problematic.

BenTheElder commented:
Having nodes as a field also means we can default adding a single nodespec when no nodes are specified.

fabriziopandini commented:
@BenTheElder I would like to discuss with you how to best address the following point

> always give the rest of the code a "Config" object to work with, which should be as much as possible simply something read from disk and desearialized with minimal mutation (just defaulting).

Currently, we are using a replica field at node level. e.g. from @neolit123 example

```
kind: KindConfig
apiVersion: kind.sigs.k8s.io/v1alpha2
nodes:
- role: control-plane
  replicas: 3
- role: worker
  replicas: 2
- role: external-load-balancer
```

However this is something the rest of the code should not work with, because the rest of the code needs something where the expected "config nodes" are derived from to the replicas number; additionally this "derive" process should take care also of assigning unique names to nodes. e.g. the list of "config nodes" for the above example should be:

```
derivedNodes:
- role: control-plane
  name: control-plane1
- role: control-plane
  name: control-plane2
- role: control-plane
  name: control-plane3
- role: worker
  name: worker1
- role: worker
  name: worker2
- role: external-load-balancer
  name: lb
```

Now that we are exposing the Config object as a public object type, it becames harder to keep the derived part "hidden" from the api-machinery that handles type conversion and defaulting, so I think it will be useful to do a quick pair review before investing to much time into this refactoring

BenTheElder commented:
I don't think we need to derive them on the object all. Config is config,
and at runtime we can have a completely distinct structure with the node
handles we've produced. When creating any node we can note the replication
number in the config.

On Thu, Dec 20, 2018, 02:34 Fabrizio Pandini <notifications@github.com
wrote:

> @BenTheElder <https://github.com/BenTheElder> I would like to discuss
> with you how to best address the following point
>
> always give the rest of the code a "Config" object to work with, which
> should be as much as possible simply something read from disk and
> desearialized with minimal mutation (just defaulting).
>
> Currently, we are using a replica field at node level. e.g. from
> @neolit123 <https://github.com/neolit123> example
>
> kind: KindConfig
> apiVersion: kind.sigs.k8s.io/v1alpha2
> nodes:
> - role: control-plane
>   replicas: 3
> - role: worker
>   replicas: 2
> - role: external-load-balancer
>
> However this is something the rest of the code should not work with,
> because the rest of the code needs something where the expected "config
> nodes" are derived from to the replicas number; additionally this "derive"
> process should take care also of assigning unique names to nodes. e.g. the
> list of "config nodes" for the above example should be:
>
> derivedNodes:
> - role: control-plane
>   name: control-plane1
> - role: control-plane
>   name: control-plane2
> - role: control-plane
>   name: control-plane3
> - role: worker
>   name: worker1
> - role: worker
>   name: worker2
> - role: external-load-balancer
>   name: lb
>
> Now that we are exposing the Config object as a public object type, it
> becames harder to keep the derived part "hidden" from the api-machinery
> that handles type conversion and defaulting, so I think it will be useful
> to do a quick pair review before investing to much time into this
> refactoring
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/kind/pull/147#issuecomment-448950860>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4Bq7cFNvgTFt3h6UQw-Z1_jhS7uS_Zks5u62fMgaJpZM4ZArB1>
> .
>

neolit123 commented:
i think we should have a discussion about this eventually.

fabriziopandini commented:
@BenTheElder @neolit123 
Moving to Config objects with nodes list as requested.
Derived fields are now sort of materialized views exiting only in the internal API; this makes easy to refactor them as soon as a better collocation is defined

BenTheElder commented:
plan is to iterate on this, @fabriziopandini's code is quite excellent and will serve as a great MVP for multi-node, thank you so much for working on this.

As discussed, we're opting to go ahead and merge this now and handle bikeshedding via follow up PR instead of infinite back and forth on the existing PRs :wink: 

Sorry for the delay, I was maybe going to cut another smaller release before putting these in, but we're just going to forge ahead now.

/hold cancel
